libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -89,11 +89,17 @@ string YulUtilFunctions::copyToMemoryFunction(bool _fromCalldata)
 			return Whiskers(R"(
 				function <functionName>(src, dst, length) {
 					calldatacopy(dst, src, length)
-					// clear end
-					mstore(add(dst, length), 0)
+					// clear up to 32 byte boundary
+					let endPos := add(dst, length)
+					if mod(endPos, 0x20)
+					{
+						<zeroMemory>(endPos, sub(<roundUp>(endPos), endPos))
+					}
```

Comment:
I was thinking about something like this (untested):
```yul
let spill := mod(endPos, 0x20)
if spill
{
    let mask := shr(not(0), sub(0x20, spill))
    mstore(endPos, and(mload(endPos), mask))
}
```
This is a single slot so using the zeroing function seems a bit heavyweight. You can instead load the current value of the slot, and clear the final bytes using a mask (all ones, shifted right by the amount of bytes we want to clear).

Also, I wonder if the optimizer is smart enough to replace `mod(endPos, 0x20)` with `and(endPos, 0x1f)` automatically. `mod` is more expensive than `and` (5 vs 3 gas) but also probably more readable so if the optimizer can do it, we might still want to use it.

Reply:
This is what the optimizer did:

```
function copy_calldata_to_memory(src, dst, length)
            {
                calldatacopy(dst, src, length)
                let endPos := add(dst, length)
                if addmod(dst, length, 0x20)
                {
                    calldatacopy(endPos, calldatasize(), sub(and(add(endPos, 31), not(31)), endPos))
                }
            }
```

Reply:
ok, it used `addmod(dst, length, 0x20)` so basically same thing as `mod(endPos, 0x20)` (maybe saves some stack operations under the hood though). Apparently it does not optimize it.

Yeah, let's use `and(endPos, 0x1f)` explicitly.

Reply:
I feel it's even worse, because it's doing the add operation twice now (`endPos := add(dst, length)` and `addmod(dst, length, 0x20)`).. maybe we can add a rule for that or fix the rule that did that?

Reply:
Yeah. That too.

But it does not look like adjusting rules would fix it. It look more like the optimizer should just reuse a common expression and it could not because `add(dst, length)` went into the `addmod`.

We could make it not convert `add` + `mod` to `addmod` but I suspect it's not a good idea in a general case even if it would help here :)